### ISSUE TITLE
JPEG: Use mozjpeg for smaller thumbnail files

### ISSUE THREAD

**Issue by vfrico** [photoprism/photoprism#1705]
After indexing all my collection of photos with photoprism I've discovered that the folder of thumbnails is huge: it is over 40 GB of space from a collection of around 30k photos. I know it is not *a lot*, but to me sounds a bit annoying.

I've found similar issues opened in github, such as [using webp, avif or jpeg xl formats](https://github.com/photoprism/photoprism/issues/668). After reading it, it appears to be a consensous of using JPEG XL when it is ready.

But, after looking how well this formats behaves today, a lot of [comparisons](https://eclipseo.github.io/image-comparison-web) with the "classic" jpeg format uses the mozjpeg encoder to compare with avif, webp and jpeg xl.

So, I've decided to take a random photo from my thumbnails folder and apply the mozjpeg encoder (built directly from the [repo](https://github.com/mozilla/mozjpeg)). These are the results before and after:

```
# Before the re-encoding. Total: ~1.2 MB

 4,2K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_100x100_center.jpg
 201K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_1280x1024_fit.jpg
  11K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_160x160_crop_28802a0ad104.jpg
 345K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_1920x1200_fit.jpg
 429K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_2048x2048_fit.jpg
  19K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_224x224_center.jpg
  18K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_224x224_left.jpg
  19K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_224x224_right.jpg
  100 nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_3x3_resize.png
  62K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_500x500_center.jpg
 1,8K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_50x50_center.jpg
  82K nov  6 15:24 888c4c43760d64f8e3413857e709bea4ecdf1037_720x720_fit.jpg
```

```
# After the reencoding. Total: ~564 KB

 3,1K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_100x100_center.jpg
  96K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_1280x1024_fit.jpg
 4,0K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_160x160_crop_28802a0ad104.jpg
 158K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_1920x1200_fit.jpg
 193K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_2048x2048_fit.jpg
 9,1K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_224x224_center.jpg
 8,9K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_224x224_left.jpg
 9,4K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_224x224_right.jpg
  30K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_500x500_center.jpg
 1,3K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_50x50_center.jpg
  41K nov  7 11:52 888c4c43760d64f8e3413857e709bea4ecdf1037_720x720_fit.jpg
```

The reduction in size is huge, specially on large thumbnails. This improves not only the space, but also the speed when using photoprism through the network.

**Comment by Fylax** [comment]
Can you please clarify the way you re-encoded images?

For my tests I used FileOptimizer, which is a GUI for an optimization chain (including, but not limited to, mozjpeg 4.0.3).
I requested lossless compression (without arithmetic encoding).

I tried to pass my whole thumbnails folder throught FileOptimizer, and I saved ~20MiB. I am surely happy of the result, but compared to the 9GiB of the whole folder looks pretty negligible.

**Comment by lastzero** [comment]
Would also be interesting if there's a performance overhead and if picture quality is really the same by actually looking at (many different) results.

**Comment by Fylax** [comment]
@lastzero generally speaking, mozjpeg provides better results than other encoders. I was actually surprised but the results I had.

On the other flip of the coin, it is known to be heavier in terms of overhead, compared to other encoders.

I think to remember that one of the reasons for highest overhead is the lack of SIMD operations

**Comment by lastzero** [comment]
In that case, it should be optional (if we add / support it). We have many NAS & Raspberry Pi users that may prefer larger files vs higher CPU load. Also it adds a dependency to an external C library, right? What if the library is not installed?

**Comment by Fylax** [comment]
I think requiring it as a dependency in docker should be fine.
I didn't really check (my bad) but I am assuming PhotoPrism relies on some Linux distribution for the container, right?
Chances that the distribution already provides the packages ready to use are quite high

**Comment by lastzero** [comment]
There's also a FreeBSD port, which doesn't use Docker.

**Comment by vfrico** [comment]
Hello! I have built mozjpeg following the instructions on the [BUILDING](https://github.com/mozilla/mozjpeg/blob/master/BUILDING.md) file. I'm on Ubuntu, and the only dependency I had to install was `libpng-dev`. 

This is one of the binary I've produced. This one is statically linked, so I assume you don't need to be aware of any external library: [cjpeg-static.zip](https://github.com/photoprism/photoprism/files/7493310/cjpeg-static.zip)

I haven't converted all the library by default, but only a few files. I'm not passing any argument to cjpeg, I'm using the default parameters. I am converting a folder of files, and saving them in other similar with this command:

```bash
find . -name "*.jpg" -exec /path/cjpeg-static -outfile aux/{} {}  \;
```

This time, the folder before the conversion had 73 elements and 8.9 MB. After passing mozjpeg, I have 4.0 MB

For compressing those files, these are my times. I'm on a mid-range desktop (AMD Ryzen): ` 3,11s user 0,12s system 99% cpu 3,239 total `

**Comment by lastzero** [comment]
If you're using the default parameters, then the JPEG quality might be lower? In that case, you should compare the file size with the equivalent JPEG quality.

**Comment by lastzero** [comment]
Obviously JPEGs get smaller the higher you compress.... Doesn't tell you what effect the special library has.

**Comment by vfrico** [comment]
Is any way to compare two images "scientifically"? I mean, to my eyes they are almost the same. I have to zoom in the image to notice some artifacts. To me the differences are no noticeable at "100% scale"

**Comment by lastzero** [comment]
http://fotoforensics.com/tutorial-estq.php

**Comment by Fylax** [comment]
By default, PhotoPrism has 80% quality. mozjpeg has 75%.
This is enough to produce smaller files.

Moreover, you are actually compressing already compressed files, thus further degrading quality.

To make more realistic measures about how much you can save with mozjpeg I can see two routes:
1. Take one original file and produce the same thumbnails (same in terms of pixels and quality) with mozjpeg, then compare the results
2. Take the already compressed files and re-encode them in loseless mode

The second route is the one I followed, producing negligible results.

**Comment by lastzero** [comment]
Our default JPEG quality actually is 92, so a reduction to 75 should lead to much smaller thumbs with less details in any case. Note this negatively impacts face recognition and image classification as described in our docs. Not just in theory.

**Comment by lastzero** [comment]
If you are OK with this, simply reduce JPEG quality to 75 in Advanced Settings. You probably don't need to use a different library as differences won't be as big as it looked like at first sight.

**Comment by lastzero** [comment]
If that (mostly) solves the issue, I suggest we wait for JPEG XL as there are more than enough other feature requests waiting to be implemented.

**Comment by erickguan** [comment]
@vfrico 
> Is any way to compare two images "scientifically"? I mean, to my eyes they are almost the same. I have to zoom in the image to notice some artifacts. To me the differences are no noticeable at "100% scale"

Yes. It's called [PSNR](https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio). But you need human to judge the photo quality.

**Comment by pyRobShrk** [comment]
When I upload a Live Photo, it uploads an HEIC image and a video. The server makes a full resolution jpeg Sidecar image. I imagine RAW photos would also have a jpeg Sidecar copy. It would seem the storage savings and quality of these full resolution jpegs would be a bit more impactful than thumbnails. It would be nice to have more control, and be able to re-generate these Sidecar images with different encoding settings like mozjpeg.

**Comment by alexislefebvre** [comment]
> It would be nice to have more control, and be able to re-generate these Sidecar images with different encoding settings like mozjpeg.

In theory it should be possible to move to the sidecar directory and process all the images with `mozjpeg`, by using `find`:

```bash
find ./ -name "*.jpg" -exec mozjpeg {} \;
```

But scanning images with Photoprism later may overwrite these files, it requires some testing.

Update: my bad, vfrico already shared this idea sooner https://github.com/photoprism/photoprism/issues/1705#issuecomment-962701893

---

Slightly off-topic: does someone have a Docker image with `mozjpeg` to share? I'm interested in testing it but I didn't find a way to use it through Docker, I only found outdated Docker images.

**Comment by erickguan** [comment]
> does someone have a Docker image with mozjpeg to share? I'm interested in testing it but I didn't find a way to use it through Docker, I only found outdated Docker images.

@alexislefebvre I could build you one from buildpacks if you are determined. Still, I'm with jpegxl. JPEG XL should have:
* better compression
* support for lossless

I deliberately not use mozjpeg given much more time in compression.

**Comment by alexislefebvre** [comment]
Google announced Jpegli: https://opensource.googleblog.com/2024/04/introducing-jpegli-new-jpeg-coding-library.html

TLDR:

> It provides both a fully interoperable encoder and decoder complying with the original JPEG standard and its most conventional 8-bit formalism, and API/ABI compatibility with libjpeg-turbo and MozJPEG.
> […]
> Our results show that Jpegli can compress high quality images 35% more than traditional JPEG codecs.

**Comment by lastzero** [comment]
> Google announced Jpegli: https://opensource.googleblog.com/2024/04/introducing-jpegli-new-jpeg-coding-library.html

Jpegli is functionality Google took from JPEG XL after they removed it from Chrome claiming that "nobody wanted it". So it seems there is some demand after all!? 🤡

**Comment by erickguan** [comment]
I am guessing that different teams made different work.

**Comment by lastzero** [comment]
> I am guessing that different teams made different work.

[Money Quote](https://opensource.googleblog.com/2024/04/introducing-jpegli-new-jpeg-coding-library.html#:~:text=Jpegli%20works%20by%20using%20a%20number%20of%20new%20techniques%20to%20reduce%20noise%20and%20improve%20image%20quality;%20mainly%20adaptive%20quantization%20heuristics%20from%20the%20JPEG%20XL%20reference%20implementation): *Jpegli works by using a number of new techniques to reduce noise and improve image quality; mainly adaptive quantization heuristics from the [JPEG XL](https://github.com/libjxl/libjxl) reference implementation [...]*

**Comment by leiserfg** [comment]
jpegli is actually part of jpegxl, and I also think it wold be a good idea to replace the jpeg implementation of photo prism with it, given that it has better quality and is backward compatible.

**Comment by lastzero** [comment]
If you choose `libvips` to generate thumbnails, we use their JPEG implementation, which in turn depends on how the Ubuntu/Debian team compiles it (the currently available packages do not seem to support mozjpeg, but jpegxl should be supported - feel free to ask the package maintainers to add mozjpeg support):
- https://docs.photoprism.app/release-notes/#may-23-2024

In any case, according to their documentation, enabling JPEG size and quality optimization comes with a significant performance overhead, so users with entry-level servers or NAS devices may not want to do so. This means we would need to add another set of config options:
https://github.com/photoprism/photoprism/blob/b70a360b4ab7ec66f71df564e40f3a43a5c6ee4b/internal/thumb/vips.go#L149-L154

**Comment by lastzero** [comment]
I will close this issue in favor of adding avif thumbnail support in a future release, which should not only reduce the file size but also improve the quality:

- https://github.com/photoprism/photoprism/issues/668

Hope that's ok for everyone!
